### PR TITLE
First last possible days available from grape

### PR DESCRIPTION
**Pull request by fonsecadeline**

### CODE REVIEW

File: test/lib/heuristics/periodic_functions_test.rb [hunk 1]
```diff
@@ -623,5 +639,50 @@ def test_lapse_when_only_one_working_week_day
...
+      refute s.send(:compatible_days, 'service_1', 1)
+    end
+
+    def test_can_find_only_value_per_set_function
+      vrp = TestHelper.create(VRP.periodic)
+      vrp.vehicles = TestHelper.expand_vehicles(vrp)
+      s = Wrappers::PeriodicHeuristic.new(vrp)
+
+      assert s.send(:can_find_only_value_per_set, [[[0, 4], [2, 3]], [[0, 4], [2, 3]]])
+      refute s.send(:can_find_only_value_per_set, [[[0, 4], [2, 3]], [[0, 4], [2, 3]], [[0, 4], [2, 3]]])
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply test/lib/heuristics/periodic_functions_test.rb:676]: Maybe the message is not clear : in the second case we provide 3 sets, containing 2 arrays. 
The function expects 3 sets in return, each one containing exactly one array such that no array in a set is in another set.

[[[0, 4], [2, 3]], [[0, 4], [2, 3]]] can return set1 = [[0,4]], set2 = [[2,3]]
[[[0, 4], [2, 3]], [[0, 4], [2, 3]], [[0, 4], [2, 3]]] can not find an array to fill third set.

File: test/lib/heuristics/periodic_functions_test.rb [hunk 2]
```diff
@@ -623,5 +639,50 @@ def test_lapse_when_only_one_working_week_day
...
+      vrp.services.first.first_possible_days = [0, 1, 2, 3]
+      vrp.services.first.last_possible_days = [4, 2, 3, 8]
+      assert s.send(:day_in_possible_interval, 'service_1', 0)
+      s.instance_variable_get(:@services_data)['service_1'][:used_days] = [0]
+      refute s.send(:day_in_possible_interval, 'service_1', 0) # only one visit can be assigned to a given day / in a given set
+      s.instance_variable_get(:@services_data)['service_1'][:used_days] = [0, 1, 2]
+      refute s.send(:day_in_possible_interval, 'service_1', 2)
+      assert s.send(:day_in_possible_interval, 'service_1', 4)
+      s.instance_variable_get(:@services_data)['service_1'][:used_days] = [0, 1, 2, 4]
+      refute s.send(:day_in_possible_interval, 'service_1', 7)
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply test/lib/heuristics/periodic_functions_test.rb:659]: Exactly :)

File: test/lib/heuristics/periodic_functions_test.rb [hunk 3]
```diff
@@ -623,5 +639,50 @@ def test_lapse_when_only_one_working_week_day
...
+
+    def test_day_in_possible_interval
+      vrp = TestHelper.create(VRP.periodic)
+      vrp.services.first.visits_number = 4
+      vrp.schedule_range_indices[:end] = 40
+      vrp.vehicles = TestHelper.expand_vehicles(vrp)
+      s = Wrappers::PeriodicHeuristic.new(vrp)
+
+      vrp.services.first.first_possible_days = [0, 1, 2, 3]
+      vrp.services.first.last_possible_days = [4, 2, 3, 8]
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply test/lib/heuristics/periodic_functions_test.rb:651]: Yes, these are not a representative first/last_possible_days but I wanted to make sure this function is able to deal with non standard cases.

File: models/vrp.rb [hunk 4]
```diff
@@ -452,6 +452,30 @@ def self.deduce_unavailable_days(hash, element, start_index, end_index, type)
...
+
+      element[:first_possible_days] = element[:first_possible_day_indices].to_a.slice(0..(element[:visits_number] || 1) - 1)
+      element[:last_possible_days] = element[:last_possible_day_indices].to_a.slice(0..element[:visits_number] || 1 - 1)
+
+      %i[first_possible_day_indices first_possible_dates last_possible_day_indices last_possible_dates].each{ |k|
+        element.delete(k)
+      }
+    end
+
+    def self.convert_possible_dates_into_indices(element, hash, start_index)
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply models/vrp.rb:466]: In fact this is called in _generate_schedule_indices_from_date_ function, it is an addition in this original logic.

File: models/vrp.rb [hunk 5]
```diff
@@ -452,6 +452,30 @@ def self.deduce_unavailable_days(hash, element, start_index, end_index, type)
...
+      return unless hash[:configuration][:schedule][:range_date]
+
+      start_value = hash[:configuration][:schedule][:range_date][:start].to_date.ajd.to_i - start_index
+      element[:first_possible_day_indices] ||= element[:first_possible_dates].to_a.collect{ |date|
+        date.to_date.ajd.to_i - start_value
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply models/vrp.rb:471]: This corresponds to astronomical Julian day. I used same logic as we originally had. We should discuss it with @braktar, you can create a ticket if you think we should do this differently.

File: lib/interpreters/periodic_visits.rb [hunk 6]
```diff
@@ -374,33 +374,34 @@ def associate_skills(new_vehicle, vehicle_day_index)
...
+          computed_first_possible_days += [day]
+          nb_services_seen += 1
+          day = [day + (service.minimum_lapse || 1), service.first_possible_days[nb_services_seen]].compact.max
         end
+        service.first_possible_days = computed_first_possible_days
 
         # last possible day
-        day = @schedule_end
         nb_services_seen = 0
+        day = [service.last_possible_days[service.visits_number - nb_services_seen - 1], @schedule_end].compact.min
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply lib/interpreters/periodic_visits.rb:394]: Here we can definetely do it with _service.last_possible_days.last_ but I thought it would be nicer to have same definition at initialization and in while loop. Maybe it was a mistake.

File: lib/interpreters/periodic_visits.rb [hunk 7]
```diff
@@ -374,33 +374,34 @@ def associate_skills(new_vehicle, vehicle_day_index)
     def compute_possible_days(vrp)
       # for each of this service's visits, computes first and last possible day to be assigned
       vrp.services.each{ |service|
-        day = @schedule_start
         nb_services_seen = 0
+        day = [service.first_possible_days[nb_services_seen], @schedule_start].compact.max
```
[the remark this answers is not in the document]
Comment (fonsecadeline) [reply lib/interpreters/periodic_visits.rb:378]: Similar reason as previous but I can change it.

File: test/lib/heuristics/periodic_functions_test.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (senhalil) [reply test/lib/heuristics/periodic_functions_test.rb:678]: ooh! I guess I did get it now. Let me rephrase it and you tell me if I understood it correctly.

Is it possible to select a subset (array) from each set (of sets) such that all selected subsets are different from each other?

If this is the case then I suggest "checks if it is possible to select an array from each set (of sets) such that all selected arrays are different from each other?" for the explanation and `can_find_a_distinct_array_per_set(sets)` for the function name, if it is okay for you too.

File: models/vrp.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (senhalil) [reply models/vrp.rb:481]: > This corresponds to astronomical Julian day. 

I already check the description of that function but I couldn't understand the **significance** of Astronomical Julian day that's why I asked if you could elaborate. What does it do and why not any other functions instead ( `ld` or `jd` for example, which already return integers in the first place, instead of a rational number) or why not we just directly do `(date.to_date - start_date).to_i` (`Date` class implements simple arithmetic operations like `-` and `+`) and Ruby handles the gritty details.

> I used same logic as we originally had.

Oh so we already have the same logic for the other "date -> index" conversion, is that it? If that is the case, can you extract the logic to a function so that we are sure that all such conversion are done via the same `convert_date_to_index(date, range_start)` function. So that there will be only one conversion logic.

> We should discuss it with @braktar, you can create a ticket if you think we should do this differently.

Sure, you can extract the `date.to_date.ajd.to_i - start_value` to a `convert_date_to_index` function and I will create a ticket to investigate if we need `ajd` or not. I am just afraid of messing up date logic because such little details can have unintended consequences -- for example [Twitter had crashed](https://www.theguardian.com/technology/2014/dec/29/twitter-2015-date-bug) for android users because a date formatted in ISO year format (%G) instead of usual/ordinary year format (%Y). Since I didn't understand the significance of Astronomical Julian day and since I don't know why we just didn't do `(date - date).to_i` in the first place, I thought I should ask ;)

File: lib/interpreters/periodic_visits.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (senhalil) [reply lib/interpreters/periodic_visits.rb:393]: I mean I find `.first` and `.last` more indicative about your intention or what you want to do but as you wish. You can resolve or change&resolve both.

File: models/vrp.rb [hunk 8]
```diff
@@ -462,6 +466,30 @@ def self.deduce_unavailable_days(hash, element, start_index, end_index, type)
       collect_unavaible_day_indices(element, start_index, end_index)
     end
 
+    def self.deduce_possible_days(hash, element, start_index)
+      convert_possible_dates_into_indices(element, hash, start_index)
+
+      element[:first_possible_days] = element[:first_possible_day_indices].to_a.slice(0..(element[:visits_number] || 1) - 1)
+      element[:last_possible_days] = element[:last_possible_day_indices].to_a.slice(0..element[:visits_number] || 1 - 1)
```
Comment (braktar) [inline models/vrp.rb:473]:
```suggestion
      element[:last_possible_days] = element[:last_possible_day_indices].to_a.slice(0..(element[:visits_number] || 1) - 1)
```
If we use parenthesis above, we should use it here too

**Review by braktar — APPROVED** [review]
